File: ruby/derived_class.rb
require_relative 'base_class'

class DerivedClass < BaseClass
  def initialize
    super
    @derived_var = "I'm a derived class variable"
  end

  def base_method
    super
    puts "This is an overridden method in DerivedClass"
  end

  def derived_method
    puts "This is a method from DerivedClass"
    puts "Derived var: #{@derived_var}"
    base_method
  end

  # Demonstrate Ruby's method_missing
  def method_missing(method_name, *args, &block)
    puts "You called a non-existent method: #{method_name}"
    puts "Arguments: #{args.join(', ')}"
  end

  # Demonstrate Ruby's respond_to_missing?
  def respond_to_missing?(method_name, include_private = false)
    method_name.to_s.start_with?('dynamic_') || super
  end
end

File: java/DerivedClass.java
public class DerivedClass extends BaseClass {
    private String derivedVar;

    public DerivedClass() {
        super();
        this.derivedVar = "I'm a derived class variable";
    }

    @Override
    public void baseMethod() {
        super.baseMethod();
        System.out.println("This is an overridden method in DerivedClass");
    }

    public void derivedMethod() {
        System.out.println("This is a method from DerivedClass");
        System.out.println("Derived var: " + derivedVar);
        baseMethod();
    }

    // Demonstrate lambda expression
    public void performOperation(int x, int y) {
        Operation addition = (a, b) -> a + b;
        System.out.println("Result of addition: " + addition.operate(x, y));
    }

    @FunctionalInterface
    interface Operation {
        int operate(int a, int b);
    }
}